clock_control: Added clock controller driver for Raspberry Pi Pico

soburi opened the pull request:
Reworked #52901 for adding clock controller driver. Changes from the original patches.

- Change to calculating frequency from register settings 
- Replacing SDK provided `clocks_init()`.
- Supporting GPOUT/GPIN
- Drop supporting set_rate/configure.

andrei-edward-popa committed:
drivers: watchdog: Changed how to get xtal frequency for Raspberry Pi Pico

Changed how to get xtal frequency for Raspberry Pi Pico

Signed-off-by: Andrei-Edward Popa <andrei.popa105@yahoo.com>
Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

andrei-edward-popa committed:
drivers: clock_control: Added clock driver for Raspberry Pi Pico

Added clock driver for Raspberry Pi Pico platform

Signed-off-by: Andrei-Edward Popa <andrei.popa105@yahoo.com>
Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

andrei-edward-popa committed:
drivers: spi: Changed how to get clock frequency for PL022

Changed how to get clock frequency for PL022

Signed-off-by: Andrei-Edward Popa <andrei.popa105@yahoo.com>
Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

andrei-edward-popa committed:
drivers: serial: Removed all function calls from Raspberry Pi Pico SDK

Removed all function calls from Raspberry Pi Pico SDK
Added functions for setting uart baudrate and format

Signed-off-by: Andrei-Edward Popa <andrei.popa105@yahoo.com>
Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

soburi committed:
modules: hal_rpi_pico: Remove unused drivers

Remove drivers that not been referenced.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

soburi committed:
driver: pwm: rpi_pico: Change to use clock controller

Since clock_control has been introduced, use it to obtain the frequency.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

soburi committed:
boards: arm: adafruit_kb2040: Added clock support in docs

Added clock in supported hardware features

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

soburi committed:
boards: arm: sparkfun_pro_micro_rp2040: Added clock support in docs

Added clock in supported hardware features

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

soburi committed:
drivers: adc: rpi_pico: Turn on clock and reset device on init

Turning on clock via clock controller and
resetting ADC device via reset controller on initializing.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

soburi committed:
drivers: usb: rpi_pico: Turn on clock on initializing

Turning on clock via clock controller on initializing.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

soburi committed:
drivers: misc: pio_rpi_pico: Turn on clock and reset device on init

Turning on clock via clock controller and
resetting PIO device via reset controller on initializing.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

soburi committed:
drivers: spi: pl022: Reset device on initializing

Reset the device on initializing via reset controller.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

soburi committed:
drivers: clock_control: rpi_pico: Configure GPOUT/GPIN pins

Configure GPOUT/GPIN pin for external clock in/out via GPIO.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

soburi committed:
drivers: counter: rpi_pico: Turn on clock and reset device on init

Turning on clock via clock controller and
resetting PIO device via reset controller on initializing.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

tmon-nordic reviewed (dismissed):
USB part looks fine.

yonsch commented on drivers/clock_control/clock_control_rpi_pico.c:
```diff
@@ -0,0 +1,885 @@
+
+DEVICE_DT_INST_DEFINE(0, &clock_control_rpi_pico_init, NULL, &clock_control_rpi_pico_data,
+		      &clock_control_rpi_pico_config, PRE_KERNEL_1,
+		      CONFIG_CLOCK_CONTROL_INIT_PRIORITY, &clock_control_rpi_pico_api);
```
Should the init priority really be configurable? Maybe this should just be a hard coded zero

soburi replied:
The priority is constrained strongly. But not we cannot be configured at all. We can configure it relative to other drivers' priority. And the default value will not cause trouble in usual case.
So, I leave as it is.

yonsch commented on soc/arm/rpi_pico/rp2/soc.c:
```diff
@@ -28,19 +28,6 @@ LOG_MODULE_REGISTER(soc, CONFIG_SOC_LOG_LEVEL);
 
 static int rp2040_init(void)
```
So now this function does nothing, it can be removed (with the SYS_INIT)

soburi replied:
Removed it.

yonsch reviewed (changes requested):
Looking good overall, added a few suggestions

soburi committed:
boards: arm: sparkfun_pro_micro_rp2040: Added counter support in docs

Added counter in supported hardware features.
It forgot in the commit when introducing the counter driver.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

soburi committed:
boards: arm: adafruit_kb2040: Added counter support in docs

Added counter in supported hardware features.
It forgot in the commit when introducing the counter driver.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

soburi committed:
boards: arm: rpi_pico_w: Added counter support in docs

Added counter in supported hardware features.
It forgot in the commit when introducing the counter driver.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

soburi commented on boards/arm/sparkfun_pro_micro_rp2040/sparkfun_pro_micro_rp2040.yaml:
```diff
@@ -19,3 +19,4 @@ supported:
   - pwm
   - flash
   - dma
+  - clock
```
Fixed it.

soburi commented on dts/bindings/clock/raspberrypi,pico-clocks.yaml:
```diff
@@ -0,0 +1,18 @@
+# Copyright (c) 2022 Andrei-Edward Popa
+# SPDX-License-Identifier: Apache-2.0
+
+description: Raspberry Pi Pico clock controller node
```
Changed it to `raspberrypi,pico-clock-controller`.

soburi commented on drivers/counter/counter_rpi_pico_timer.c:
```diff
@@ -161,6 +163,12 @@ static void counter_rpi_pico_irq_handle(uint32_t ch, void *arg)
 	const struct counter_rpi_pico_timer_config *config = dev->config;
+	int ret;
+
+	ret = reset_line_toggle_dt(&config->reset);
```
It is not necessary in actually because using clk_ref, but
watchdog turns on the clk_ref clock explicitly, so do same as it.

soburi commented on drivers/clock_control/Kconfig.rpi_pico:
```diff
@@ -0,0 +1,26 @@
+
+if CLOCK_CONTROL_RPI_PICO
+
+config XOSC_STARTUP_DELAY
```
It is not used.
I forgot to remove that added in experimental.
Removed it.

soburi commented on drivers/spi/spi_pl022.c:
```diff
@@ -266,13 +268,20 @@ struct spi_pl022_dma_data {
 	const uint32_t reg;
 	const uint32_t pclk;
 	const bool dma_enabled;
+#if defined(CONFIG_CLOCK_CONTROL)
```
That's better.
I fixed it.

soburi committed:
driver: spi: spi_rpi_pico_pio: Change to use clock controller

Since clock_control has been introduced, use it to obtain the frequency.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

soburi committed:
boards: arm: rpi_pico: Added clock support in documentation

Added clock in supported hardware features

Signed-off-by: Andrei-Edward Popa <andrei.popa105@yahoo.com>
Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

soburi committed:
drivers: pwm: rpi_pico: Reset device on init

Resetting PWM device via reset controller on initializing.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

soburi commented:
@yonsch 
Could you take a look, please?

soburi commented:
@yonsch 
ping

soburi commented:
rebased

ajf58 commented:
Is there anything I can do to help progress this PR?
